### PR TITLE
Add AMOs to CommDiagnostics.

### PR DESCRIPTION
**Pull request by gbtitus**
We've long wanted to include atomic operations in comm diagnostics.  Do
that here.  But param-disable printing atomic op counts by default,
because those counts are unstable due to our use of network atomics when
spin-waiting for parallel construct and on-stmt completion.

This is related to #10124, but doesn't resolve that because it doesn't enable
counting AMOs by default.

### CODE REVIEW

**Comment by gbtitus** [comment]
It would be nice if `commDiagsPrintUnstable` were a `config const` instead of a `config param`, but doing that causes a compiler error I don't understand.  Rather than my attempting to describe it, it might be best if someone else (hint, hint) reproduced that themselves.

**Comment by gbtitus** [comment]
The `commDiagsPrintUnstable` flag should also control verbose printing of "unstable" ops.

**Comment by ronawho** [comment]
Do you think it's worth counting fetching vs non-fetching ops separately?

**Comment by gbtitus** [comment]
> Do you think it's worth counting fetching vs non-fetching ops separately?

I'm on the fence whether that's too fine a partitioning or not.  One possibility (long-term rather than now, say) would be to do the counting itself with a maximally fine partitioning (fetching/nonfetching and vanilla/nb/unordered/fast etc.).  Then add one or more arguments to `getCommDiagnostics*()` to say whether and how to combine the various subcounts into the main counts in the return value, with the defaults for those arguments chosen so as to reproduce the current results.

**Comment by gbtitus** [comment]
comm=ofi full testing passes on our IB-based cluster.  With respect to comm=ugni, which is the only other comm layer that can do network atomics, it doesn't look like there's any overlap between tests that use `CommDiagnostics` and those that are tested on Cray XC, with the exception of 3 tests for which there are `.ml-*` files and which therefore would be covered by performance testing.  Those 3 all pass.

**Comment by gbtitus** [comment]
This will be easiest to look at commit-by-commit.  It's ready for re-review but I still have some testing to do.

**Comment by ronawho** [comment]
I think this looks ok, but there's a decent amount here so I'm not sure I grok all of it. How would you feel about splitting this into 2 PRs -- one that does cleanup and the other that actually adds the comm diags for AMOs?

Also maybe link to https://github.com/chapel-lang/chapel/issues/10124 (but leave it open till amo diags are on by default)

**Comment by ronawho** [comment]
> It would be nice if commDiagsPrintUnstable were a config const instead of a config param, but doing that causes a compiler error I don't understand. Rather than my attempting to describe it, it might be best if someone else (hint, hint) reproduced that themselves.

Looks to be an early order of initialization/resolution error. If you add an arbitrary `config const` to `CommDiagnostics` you'll get the same error, there's nothing special about `commDiagsPrintUnstable`. 

e.g. on master just adding `config const foo = 1;` to `CommDiagnostics` and compiling with `--print-callstack-on-error` I see:

```
$CHPL_HOME/modules/internal/String.chpl:302: error: unable to resolve type
while processing the following Chapel call chain:
  $CHPL_HOME/modules/internal/String.chpl:302: string [internal module] [compiler-generated]
  $CHPL_HOME/modules/internal/ChapelBase.chpl:1215: _command_line_cast(param s: c_string, type t, x: c_string) [internal module]
  $CHPL_HOME/modules/standard/CommDiagnostics.chpl:170: top-level module statements for CommDiagnostics
```

File: modules/standard/CommDiagnostics.chpl [hunk 1]
```diff
@@ -224,8 +240,11 @@ module CommDiagnostics
       for param i in 1..numFields(chpl_commDiagnostics) {
         const val = getField(this, i);
         if val != 0 {
-          if first then first = false; else c <~> ", ";
-          c <~> getFieldName(chpl_commDiagnostics, i) <~> " = " <~> val;
+          if getFieldName(chpl_commDiagnostics, i) != 'amo' ||
+             commDiagsPrintUnstable {
+            if first then first = false; else c <~> ", ";
+            c <~> getFieldName(chpl_commDiagnostics, i) <~> " = " <~> val;
+          }
```
Comment (ronawho) [inline modules/standard/CommDiagnostics.chpl]: Might be a little cleaner as:

```chpl
        param name = getFieldName(this.type, i);
        const val = getField(this, i);
        if val != 0 {
          if commDiagsPrintUnstable || name != 'amo' {
            if first then first = false; else c <~> ", ";
            c <~> name <~> " = " <~> val;
          }
        }
```

**Comment by gbtitus** [comment]
> Looks to be an early order of initialization/resolution error.

Yeah, that part I had kind of guessed was the case, and I knew it wasn't anything in particular about my symbol that was the problem.  I think I'll just leave it a param for now.

**Comment by gbtitus** [comment]
> How would you feel about splitting this into 2 PRs -- one that does cleanup and the other that actually adds the comm diags for AMOs?

Will do.

**Comment by gbtitus** [comment]
This PR was superseded by #12951.

**Pull request closed without merging**